infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -0,0 +1,514 @@
+            log.info("Start as Sink (receiver) on node {} ", localNodeDescriptor);
+        } else {
+            log.error("Log Replication not started on this cluster. Leader node {} belongs to cluster with {} role.",
+                    localEndpoint, localNodeDescriptor.getRoleType());
+        }
+    }
+
+    private void updateTopologyConfigId(boolean active) {
+        // Required only on topology changes
+        logReplicationServer.getSinkManager().updateTopologyConfigId(active, topologyDescriptor.getTopologyConfigId());
+
+        log.debug("Persist new topologyConfigId {}, status={}", topologyDescriptor.getTopologyConfigId(),
+                localNodeDescriptor.getRoleType());
+    }
+
+    /**
+     * Stop ongoing Log Replication
+     */
+    private void stopLogReplication() {
+        if (localNodeDescriptor.isLeader() && localNodeDescriptor.getRoleType() == ClusterRole.ACTIVE) {
+            replicationManager.stop();
+        }
+    }
+
+    /**
+     * Process lock acquisition event
+     */
+    public void processLockAcquire() {
+        log.info("Process lock acquire event");
+
+        logReplicationServer.setLeadership(true);
+
+        // TODO(Gabriela): confirm that start does not affect ongoing replication if it is called again..
+        if (!localNodeDescriptor.isLeader()) {
+            // leader transition from false to true, start log replication.
+            localNodeDescriptor.setLeader(true);
+            startLogReplication();
+        }
+    }
+
+    /**
+     * Process lock release event
+     *
+     * Set leadership metadata and stop log replication in the event of leadership loss
+     */
+    public void processLockRelease() {
+        logReplicationServer.setLeadership(false);
+
+        if (localNodeDescriptor.isLeader()) {
+            stopLogReplication();
+            localNodeDescriptor.setLeader(false);
+        }
+    }
+
+
+    public void processSiteFlip(TopologyDescriptor newConfig) {
+        // TODO(Nan): Check standby to active and active to standby...
+        stopLogReplication();
+        //TODO pankti: read the configuration again and refresh the LogReplicationConfig object
+        replicationManager.setTopologyDescriptor(newConfig);
+        boolean activeCluster = localNodeDescriptor.getRoleType() == ClusterRole.ACTIVE;
+        updateTopologyConfigId(activeCluster);
+        startLogReplication();
+    }
+
+    public void processSiteChangeNotification(DiscoveryServiceEvent event) {
+        // Stale notification, skip
+        if (event.getTopologyConfig().getTopologyConfigID() < getReplicationManager().getTopologyDescriptor().getTopologyConfigId()) {
+            log.debug("Stale Topology Change Notification, current={}, received={}", topologyDescriptor.getTopologyConfigId(), event.getTopologyConfig());
+            return;
+        }
+
+        TopologyDescriptor newConfig = new TopologyDescriptor(clusterManagerAdapter.fetchTopology());
```

Comment:
Do we assume here that this call will always succeed?

Reply:
No, I have changed this. It can certainly fail.